src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.Analysis.cs

```diff
@@ -533,28 +533,42 @@ private BoundNode VisitLambdaOrFunction(IBoundLambdaOrFunction node)
 
             private void ReferenceVariable(SyntaxNode syntax, Symbol symbol)
             {
-                var localSymbol = symbol as LocalSymbol;
-                if ((object)localSymbol != null && localSymbol.IsConst)
+                if (symbol is LocalSymbol localSymbol && localSymbol.IsConst)
                 {
                     // "constant variables" need not be captured
                     return;
                 }
 
-                // using generic MethodSymbol here and not LambdaSymbol because of local functions
-                MethodSymbol lambda = _currentParent as MethodSymbol;
                 // "symbol == lambda" could happen if we're recursive
-                if ((object)lambda != null && symbol != lambda && symbol.ContainingSymbol != lambda)
+                if (_currentParent is MethodSymbol lambda && symbol != lambda && symbol.ContainingSymbol != lambda)
                 {
                     CapturedVariables.Add(symbol, syntax);
 
                     // mark the variable as captured in each enclosing lambda up to the variable's point of declaration.
-                    for (; (object)lambda != null && symbol != lambda && symbol.ContainingSymbol != lambda; lambda = lambda.ContainingSymbol as MethodSymbol)
+                    while ((object)lambda != null &&
+                           IsLambdaOrLocalFunction(lambda) &&
+                           symbol != lambda &&
+                           symbol.ContainingSymbol != lambda)
                     {
                         CapturedVariablesByLambda.Add(lambda, symbol);
+                        lambda = lambda.ContainingSymbol as MethodSymbol;
                     }
                 }
             }
 
+            private static bool IsLambdaOrLocalFunction(MethodSymbol method)
```

Comment:
Do we need to check for that? Would the `symbol != lambda` condition imply that we see a lambda or a local function?